server/src/main/java/org/opensearch/action/termvectors/TransportTermVectorsAction.java

```diff
@@ -87,15 +88,24 @@ public TransportTermVectorsAction(
 
     @Override
     protected ShardIterator shards(ClusterState state, InternalRequest request) {
+
+        String preference = request.request().preference;
+        // For a real time request on a seg rep index, use primary shard as the preferred query shard.
+        if (request.request().realtime()
+            && preference == null
+            && state.getMetadata().isSegmentReplicationEnabled(request.concreteIndex())) {
+            preference = Preference.PRIMARY.type();
+        }
+
         if (request.request().doc() != null && request.request().routing() == null) {
             // artificial document without routing specified, ignore its "id" and use either random shard or according to preference
             GroupShardsIterator<ShardIterator> groupShardsIter = clusterService.operationRouting()
-                .searchShards(state, new String[] { request.concreteIndex() }, null, request.request().preference());
+                .searchShards(state, new String[] { request.concreteIndex() }, null, preference);
```

Comment:
@Rishikesh1159 I don't think the primary override here is necessary. In this case, an "artificial" document has been supplied and the user is asking to compute term vectors as if this was a real document. I don't think the "real time" parameter is applicable here since it doesn't need to look up any index data.

@msfroh Does that sound right to you?

Reply:
That does sound right to me. 

I think the override condition should be:
```
        if (request.request().realtime()
            && preference == null
            && state.getMetadata().isSegmentReplicationEnabled(request.concreteIndex())
            && request.request().doc() == null) {
```

Reply:
Interestingly the code seems to [unconditionally attempt to fetch the document](https://github.com/opensearch-project/OpenSearch/blob/aca2e9de7daaadee77c33e21ff5215c3f1e8600f/server/src/main/java/org/opensearch/index/termvectors/TermVectorsService.java#L114-L118) even if an artificial document is supplied. Not sure if that complicates things here.

Reply:
I think that's probably worth fixing -- from the looks of it, that `GetResult` is ignored, which seems a little wasteful.

That said, I don't think it *hurts* anything. In particular, if the `doc` parameter is specified, then it GETs a document with ID based on an autoincrementing (not really random) integer: https://github.com/opensearch-project/OpenSearch/blob/aca2e9de7daaadee77c33e21ff5215c3f1e8600f/server/src/main/java/org/opensearch/action/termvectors/TermVectorsRequest.java#L315-L323

(`generateRandomInteger` is always true on the REST parsing code path.)

Reply:
Still, there is no value in routing to a primary shard if an artificial doc is specified.